### ISSUE TITLE
Ahoy with bot_detection_version 2 detects a lot of real android devices as bots

### ISSUE THREAD

**Issue by ts-3156** [ankane/ahoy#400]
Hi,

Thanks for this useful gem!

Current default bot detection (bot_detection_version 2) detects a lot of real android devices as bots. I know it's not ahoy's fault. However, it does not seem friendly as a starting point for people who start using this gem. How about changing the default bot detection configuration?

## Steps to reproduce

When `device_type.nil?` is true, ahoy detects the user agent is a bot.

```ruby
# https://github.com/ankane/ahoy/blob/master/lib/ahoy/base_store.rb#L48-L68
def bot?
  unless defined?(@bot)
    @bot = begin
      if request
        if Ahoy.user_agent_parser == :device_detector
          if Ahoy.bot_detection_version == 2
            DeviceDetector.new(request.user_agent).device_type.nil? # <- Here!
          else
            DeviceDetector.new(request.user_agent).bot?
          end
        else
          Browser.new(request.user_agent).bot?
        end
      else
        false
      end
    end
  end

  @bot
end

```

`DeviceDetector#device_type` returns nil when I give a parser a user agent of Pixel3.

```ruby
# This is a real user agent of my smartphone.
pixel3 = 'Mozilla/5.0 (Linux; Android 9.0.99) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36'
detector = DeviceDetector.new(pixel3)

detector.device_type.nil?
=> true

detector.bot?
=> false

detector.name
=> "Chrome Mobile"

detector.os_name
=> "Android"
```

In my rails application with ahoy, user agents of the below is detected as bots.

```
# Sony
Mozilla/5.0 (Linux; Android 7.0; SOV32) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36

# Sony
Mozilla/5.0 (Linux; Android 7.0; SO-01J) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36

# Huawei
Mozilla/5.0 (Linux; Android 7.0; WAS-LX2J) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36

# Huawei
Mozilla/5.0 (Linux; Android 9; LYA-L09) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36

# Sharp
Mozilla/5.0 (Linux; Android 8.0.0; SHV40) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36

# Fujitsu
Mozilla/5.0 (Linux; Android 7.1.1; F-04H Build/V11R045B; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/73.0.3683.90 Safari/537.36
```

## Solution

I deal with this issue by using other customized user agent parser as a transitory measure. However, I know this is not a fundamental solution.

**Comment by ankane** [comment]
Hey @ts-3156, I'm open to suggestions on how to improve the bot detection logic. The latest logic was a result of the earlier logic allowing many bots through.

**Comment by ts-3156** [comment]
Hi,

> The latest logic was a result of the earlier logic allowing many bots through.

I feel for you... How about writing "Current bot detection is intentionally false-positive. Change the logic as needed" to README.md as a note?

I think if the current logic is intentional, changing the logic several times is a kind of pointless. However, writing a note may be kind for beginners. 

I'm sorry I can't be of any help! Please feel free to close this issue when you think this discussion is finished. Have a good weekend.

**Comment by ankane** [comment]
Hey @ts-3156, I appreciate the suggestion, but not sure it adds much to the readme. If you have time, I suggest submitting the issue to the device_detector gem.

**Comment by dkam** [comment]
Thanks so much for writing and releasing this gem (and all your other gems!).  I'm a new user and I'm finding it super useful. 

It looks like the inverse is also happening, bot_detect_version 2 is classifying actual bots as not bots.

> When device_type.nil? is true, ahoy detects the user agent is a bot. 

When device_type.nil? is false, ahoy does not detects the user agent as a bot.

The following example User Agent string is detected as device 'smartphone' and so is not classified as a bot. 

```ruby
ua = "Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.96 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
DeviceDetector.new(ua).device_type
=> "smartphone"
DeviceDetector.new(ua).device_type.nil?
=> false
DeviceDetector.new(ua).bot?
=> true
```

I'll try using ```Ahoy.bot_detection_version = 2```

**Comment by ankane** [comment]
@dkam Didn't realize bots could have a device type. Pushed a fix in c35a0de526b0231b1f5ea0eb8c222749be44bb88.

**Comment by ankane** [comment]
@ts-3156 Just fyi, pushed an update that fixes Android 9 logic, which will go out with Ahoy 3.0.
